server/mqtt.go

```diff
@@ -952,7 +952,20 @@ func (c *client) mqttParse(buf []byte) error {
 }
 
 func (c *client) mqttTraceMsg(msg []byte) {
-	maxTrace := c.srv.getOpts().MaxTracedMsgLen
+	opts := c.srv.getOpts()
+	maxTrace := opts.MaxTracedMsgLen
+	headersOnly := opts.TraceHeaders
+
+	// If TraceHeaders is enabled, extract only the header portion of the msg.
+	if headersOnly {
+		msg, _ = c.msgParts(msg)
+	}
+
+	// Do not emit a log line for empty zero-length payloads.
+	if len(msg) == 0 {
```

Comment:
You could however use the same fix here and check for `len(msg)-LEN_CR_LF` to decide that you want to not trace the payload.

Reply:
Actually not even, since MQTT don't have this trailing `\r\n`, this is the reason I had to make a "copy" of the function to make it special for MQTT clients. So I would completely undo the changes in that function.